Fix conflicting Serverless scaling options (#15576)
kyma-project/kyma: components/function-controller/design/scaling_functions.md

# Serverless Functions Scaling Modes

## Summary

> Initially, the Serverless API supported the `spec.ScaleConfig` field only. The work flow was as follow:

Reviewer: ```suggestion
Initially, the Serverless API supported the `spec.ScaleConfig` field only. The workflow was as follows:
```

> Initially, the Serverless API supported the `spec.ScaleConfig` field only. The work flow was as follow:

Reviewer: I changed lines 4-8 to past tense, because we're talking about the previous workflow, correct?

- Function resources were defaulted to min/max replicas = 1

> - If the user sets a different scaling config, the function controller will create an HPA resources with the user defined min/max values. 

Reviewer: ```suggestion
- If you set a different scaling config, Function Controller created HPA resources with the user-defined min/max values. 
```

> - The HPA resource target ref will be the Function runtime deployment.

Reviewer: ```suggestion
- The HPA resource target ref was the Function runtime deployment.
```

> - The Function controller will not enforce the runtime deployment `spec.Replicas` any more and it will be handled by the HPA resources.

Reviewer: ```suggestion
- The Function Controller did not enforce the runtime deployment `spec.Replicas` anymore and it was handled by the HPA resources.
```

> Later, we extended the Serverless API to include a `Scale` subresource, which allows us to directly scale the function resources through the Kubernetes API.

Reviewer: ```suggestion
Later, the Serverless API was extended by the `Scale` subresource, which allows for direct scaling of the Function resources through the Kubernetes API.
```

However, there are some implementation conflicts between the two features. This is a design and an implementation plan to unify the UX while using Functions as scaled resources.

### Goals
- Support Function scale subresource and `spec.ScaleConfig` without conflicts.
- Provide frictionless UX for the feature.

## Proposal
Describe and implement two different scaling configuration. Both configurations already work to some extent. The point here is have an ergonomic UX and flow.

### External scaling configuration
This is managed and configured using `spec.Replicas`. It describes the scale subresource use case. It supports:

> - Manually scaling the Function up/down through the API.

Reviewer: ```suggestion
- Manual scaling of the Function up and down through the API.
```

- Configuring an HPA resource with the Function resources as a target.

> - Using an external scaler like KEDA.

Reviewer: ```suggestion
- Using an external scaler like [KEDA](https://keda.sh/).
```

### Built-in scaling configuration

> This is managed and enabled by setting `spec.ScaleConfig`. It should be configurable using Busola and it provide the most basic scaling configuration for the function.

Reviewer: ```suggestion
This is managed and enabled by setting `spec.ScaleConfig`. It is configurable using Busola and it provides the most basic scaling configuration for the Function.
```

> This configuration will be disabled by removing `spec.ScaleConfig`. Busola UI can be extended to allow users to add/remove `spec.ScaleConfig` to manage built-in scaling with minimal effort to the user.

Reviewer: ```suggestion
This configuration is disabled by removing `spec.ScaleConfig`. The Busola UI can be extended to allow you to add or remove `spec.ScaleConfig` to manage built-in scaling with minimal effort.
```

## Implementation details

- The controller should support and accept both `spec.Replicas` and `spec.ScaleConfig`. Current validation rule to block this will be removed.

> - `spec.Replcas` is the only source of truth for scaling the function/runtime deployment.

Reviewer: ```suggestion
- `spec.Replicas` is the only source of truth for scaling the Function/runtime deployment.
```

- `spec.ScaleConfig` is only used to configure the controller internal HPA.
- The internal HPA is removed if `spec.ScaleConfig == nil`

> - The HPA resource created by the controller will still target the Function resources.

Reviewer: ```suggestion
- The HPA resource created by the controller still targets the Function resources.
```

> - The current Function status update logic should be fixed to reflect the function current scale. 

Reviewer: ```suggestion
- The current Function status update logic must be fixed to reflect the Function's current scale. 
```

# Serverless Functions Scaling Modes

## Summary
Initially, the Serverless API supported the `spec.ScaleConfig` field only. The work flow was as follow:
- Function resources were defaulted to min/max replicas = 1
- If the user sets a different scaling config, the function controller will create an HPA resources with the user defined min/max values. 

> - The HPA resource target ref will be the Function runtime deployment.

Reviewer 2: We can target function as it's scalable and update the deployment `replicas` using function `spec`

Author: We actually discussed this. I didn't like initially, it makes sense because this why we avoid having a conflict with other user-created HPA resources to scale the function externally. Also mapping the work flow would be confusing: function.Spec.ScaleConfig.min/max -> Internal HPA resource - > function.Spec.Replicas - > runtime deployment. spec.Replicas :/

Reviewer 2: This will hide user misconfiguration of scaling.

Author: Can you give an example?

Reviewer 2: If user set `scaleConfig` he is aware that it enables scaling. In this situation the user shouldn't create any external scaler. As it can introduce bugs. 
Additionally our CR starting behaving strangely because we will set `scaleConfig` to the same values as `replicas` and will wipe user configuration.
The scaling topic should be well documented but it's in other task.

The same situation: user shouldn't create two external scalers as it can undefined behavior.

Adding such behaviour we hide user misconfiguration of scaling and changing the configuration. 

I think that the workflow you mentioned simplify code which is responsible for managing deployments.
